File: intro.rb
#
# See https://learnxinyminutes.com/docs/ruby/
#

puts("1. hello world") # print a "hello world" statement to stdout
puts("2. hello world"); # use of the semicolon to end a statement is optional (rather, ruby looks for the end of line
puts("3. hello"); puts("world"); # puts injects a special newline character (\n)

# single vs. double-quotes
puts('\t4. hello world\n');
puts("\t5. hello world\n");

# braces may be omitted (braces hint at puts being a method, of a
# hidden, global / main object)
puts "6. hello world"

# variables: naming convention

first_name = "russ"
lastName = "sobernig"
_full_name = first_name + " " + lastName # string concatenation

puts(first_name)
puts(lastName)
puts(_full_name)

# multi-line expressions

x = 2 + 3 +
  4 + 5

puts(x)

_full_name = first_name + " " + \
  lastName

puts(_full_name)

# CONSTANTS

POUNDS_PER_KILOGRAM = 2.2
StopToken = 'stop'

puts("StopToken -> " + StopToken)

# redefining a CONSTANT is prohibited
# StopToken = 'end' # throws 'already initialized constant StopToken'

# Structured and non-structured (primitive) datatypes are all
# represented as objects

puts(7.class)
puts("hello world".class)

seven_as_string = 7.to_s # to-string conversion
y = seven_as_string.class.to_s + " -> " + seven_as_string
puts(y)

# basic control structures

# if-else-end

if 1 
  puts("7. hello world!");
end

if 0
  puts("8. hello world!");
end

# expressions

weight = 50

if(weight < 1)
  puts('very light')
elsif(weight < 10)
  puts('a bit of a load')
elsif weight < 100 # braces are, again, optional
  puts('heavy')
else
  puts('way too heavy')
end

# variable references + substitution
# in-brace vs. out-brace substitution

x = "someVar"
puts("x -> " + x) # out-brace
puts("x -> #{x}") # in-brace

puts("arbitrary expressions -> #{(3*4).to_s + "---" + first_name}") # in-brace

# loops

# 1. while
i = 0
while(i < 4)
  puts("i = #{i}")
  i = i + 1; # i += 1
end

# 2. for

some_array = ['first' , 'second' , 3 ]
for element in some_array
  puts("#{element}")
end
             
# iterators (handy!)

some_array.each do |element|
  puts(element);
end

# symbols (immutable strings, used as identifiers (object identifiers,
# object members)

# container data structures: arrays

x = []
y = Array.new()
some_array = ['first' , 'second' , 3 ]

puts(some_array[0])
puts(some_array[1])
puts(some_array[2])

puts(some_array.length()) # length introspection
puts(some_array.size()) # size introspection

some_array[2] = "three" # positional assigment

some_array[3] = 4 # positional append
some_array << 5 # relative append

# containder data structures: hashes

h = {}
h['first_name'] = ' Albert'
h['last_name'] = ' Einstein'
h['first_name']     # Is ' Albert'
h['last_name']      # Is Einstein

h = {'first_name' => ' Albert' , 'last_name' => 'Einstein'}

h.each do |key,value|
  puts(key + " -> " + value)
end

# a simple function (call)

def square(x)
	return x*x
end

puts square(2); # returns/ prints '4'

File: intro.py
#
# See https://learnxinyminutes.com/docs/python3/
#
print("1. hello world") # print a "hello world" statement to stdout
print("2. hello world"); # use of the semicolon to end a statement is optional (rather, ruby looks for the end of line
print("3a. hello"); print("world"); # puts injects a special newline character (end="\n")
print("3b. hello", end=" "); print("world");

# single vs. double-quotes: no difference when it comes to control characters
print('\t4. hello world\n');
print("\t5. hello world\n");

# to escape control characters from being interpreted, prefix the string using "r" or "R" as for "raw"
print(r"\t5. hello world\n");

# BEWARE: braces may *not* be omitted (as in Python 2)
# print "6. hello world"; # WRONG!

# All names in Python are *case-sensitive*
first_name = "russ"; # preferred style by convention: lower case and underscore as separator
lastName = "sobernig"; # mixedCase is possible, but should be avoided

_full_name = first_name + " " + lastName # string concatenation

print(first_name)
print(lastName)

print(_full_name)
print(first_name, lastName, sep=" ")

#
# CONSTANTS are written (by convention) in upper case; but there are
# no semantics attached to them
#

POUNDS_PER_KILOGRAM = 2.2

#
# Multi-line strings using triple-quote notation:
#
x = """This is a very long string, 
containing commas, that I split up
into multiple line"""

print(x)

#
# Multi-line strings using brace (tuple) notation:
#

x = ("This is a very long string "
    "containing commas, that I split up "
    "into multiple line")

print(x)

#
# Multi-line strings using break-line notation:
#

x = "This is a very long string " \
  "containing commas, that I split up " \
  "into multiple line"

print(x)
  
# Multi-line expressions using break lines

x = 2 + 3 + \
  4 + 5

print(x)

_full_name = first_name + " " + \
  lastName

print(_full_name)


# Structured and non-structured (primitive) datatypes are all
# represented as objects

print(type(7))
print(type("hello world"))

seven_as_string = str(7) # to-string conversion
print(type(seven_as_string))

print(type(int(seven_as_string))) # string-to-number conversion


# basic control structures

# if-else-end

if 1: 
  print("7. hello world!");

if True: 
  print("7. hello world!");

if 0:
  print("8. hello world!");

if False:
  print("8. hello world!");

# expressions

weight = 50

if weight < 1:
  print('very light')
elif weight < 10:
  print('a bit of a load')
elif weight < 100:
  print('heavy')
else:
  print('way too heavy')

# loops

# 1. while
i = 0
while i < 4:
  print("i = " + str(i))
  i = i + 1; # i += 1

# 2. for

some_list = ['first' , 'second' , 3 ]
for element in some_list:
  print(element)
  
# container data structures: lists

x = [] # empty list
some_list = ['first' , 'second' , 3 ]

print(some_list)
print(some_list[0])
print(some_list[1])
print(some_list[2])
print(some_list[-1]) # backward positional lookup

print(len(some_list)) # length introspection

some_list[2] = "three" # positional assignment

# some_list[3] = 4 # WRONG: positional append
some_list.append(5) # relative append
some_list.extend([6, 7, "eight"]) # relative extend, each element is added
some_list.append([6, 7, "eight"]) # append vs. extend, append argument is added as one element

print(some_list)

del some_list[1]; # remove a list element; by index
some_list.remove('first'); # remove a list element; by value

# Python has alternative list-like data structures:
# - tuples: immutable list
# - set: unordered bags of elements

# container data structures: dictionaries

h = {}
h['first_name'] = ' Albert'
h['last_name'] = ' Einstein'
h['first_name']     # Is ' Albert'
h['last_name']      # Is Einstein

h = {'first_name' : ' Albert' , 'last_name' : 'Einstein'}

print(h)

# TODO: Implement a square routine

def square(x):
    return x * x

print(square(4));
print(square(x = 4));

raise SystemExit
